GoMeet/src/main/java/com/google/sps/data/AutoProposeTimes.java

```diff
@@ -0,0 +1,195 @@
+    * call APIs from.
+    * @param apiKey The API Key to use the Google Calendar API.
+    * @param calId This list of calendar IDs of calendars to automatically generate
+    * meeting times from.
+    * @param startTime The start of the period in which to search for meeting times.
+    * @param endTime The end of the period in which search for meeting times.
+    * @param duration The duration of the meeting, in milliseconds.
+    */
+  public AutoProposeTimes(Calendar service, String apiKey, ArrayList<String> calId, 
+      Date startTime, Date endTime, int duration) {
+    this.service = service;
+    this.apiKey = apiKey;
+    this.calendarId = calId;
+    this.startTime = new DateTime(startTime, TimeZone.getTimeZone(TIMEZONE));
+    this.endTime = new DateTime(endTime, TimeZone.getTimeZone(TIMEZONE));
+    this.meetingDuration = duration;
+  }
+
+  /**
+   * Used for sorting TimePeriods from earliest start time
+   * to latest start time.
+   */
+  static class TimePeriodComparator implements Comparator<TimePeriod> {
+    public int compare(TimePeriod a, TimePeriod b) { 
+      // Compare the number of milliseconds since the Unix epoch of 
+      // the TimePeriod starts.
+      return (int) (a.getStart().getValue() - b.getStart().getValue()); 
+    } 
+  }
+
+  /**
+   * Generates a List of TimePeriod representing free times with a duration of 
+   * at least the value of 'meetingDuration'.
+   * @return A List of TimePeriod that are free periods where a
+   * meeting may be scheduled.
+   */
+  public List<TimePeriod> proposeTimes() 
+      throws IOException, GoogleJsonResponseException, GeneralSecurityException {
+    // Gather freebusy information for each of the calendars.
+    List<TimePeriod> busyPeriods = new ArrayList<TimePeriod>();
+    List<TimePeriod> freePeriods = new ArrayList<TimePeriod>();
+    for (int i = 0; i < calendarId.size(); i++) {
+      busyPeriods.addAll(freebusyRequest(calendarId.get(i)));
+    }
+
+    // If there are no busyPeriods, add a free time that is the whole period, and return
+    if (busyPeriods.size() == 0) {
+      freePeriods.add(newTimePeriod(this.startTime, this.endTime));
+      return freePeriods;
+    }
+
+    // Sort all the busy periods by start time: O(Nlog(N))
+    Collections.sort(busyPeriods, new TimePeriodComparator());
+
+    long durationStartOfDay;
+    // Add the free time block from startTime to first busy period, if any.
+    DateTime firstBusyPeriod = busyPeriods.get(0).getStart();
+    durationStartOfDay = firstBusyPeriod.getValue() - this.startTime.getValue();
+    if (durationStartOfDay >= this.meetingDuration) {
+      freePeriods.add(newTimePeriod(this.startTime, firstBusyPeriod));
+    }
+
+    // Generate the free periods: O(N)
+    int i = 0;
+    TimePeriod current = busyPeriods.get(i);
+    DateTime lastBusyPeriodEnd = current.getEnd();
+
+    while (i < busyPeriods.size()) {
+      int j = i + 1;
+      while (j < busyPeriods.size() 
```

Comment:
Another way you could do this is by sorting all the start and end times independently, traverse the list from start to end, and keep a counter for how many events you're "currently in". When the counter is 0, it means you've got some free time. e.g.,

0800 - start of day; numCurrentEvents = 0
0900 - start meeting 1; numCurrentEvents = 1 --> append(new freeTime(0800, 0900))
1000 - start meeting 2; numCurrentEvents = 2
1100 - end meeting 1; numCurrentEvents = 1
1200 - end meeting 2; numCurrentEvents = 0
1700 - end of day --> append(new freeTime(1200, 1700))

If you treat start of day as the "end" of an event, and the end of the day as the "start" of an event, your beginning-of-day and end-of-day edge cases are also eliminated.

Reply:
note this is optional, you already have something working.

Reply:
This is a smart alternative - I didn't think to include the start and end times like that 😅 For the times sake and in case I break anything, I'm going to leave what I have, but thank you for the feedback!